File: contrib/profile/hrtime/hrtime_c.pl
:- module(hrtime_c,[
	main/0,
	get_hrtimef/2,
	get_hrvtimef/2,
	get_hrutimef/2,
	get_hrstimef/2,
	get_current_hrtimef/1,

	get_hrtime_selff/1,
	get_hrvtime_selff/1,
	get_hrutime_selff/1,
	get_hrstime_selff/1,

	hrtime_initl/1,
	get_hrtime_structl/3,
	free_hrtime_structl/2],[assertions,foreign_interface]).

% the f means that the result is double
:- true pred get_hrtimef(in(Hr), go(Dest)) :: address * num + (foreign).
:- true pred get_hrvtimef(in(Hr), go(Dest)) :: address * num + (foreign).
:- true pred get_hrutimef(in(Hr), go(Dest)) :: address * num + (foreign).
:- true pred get_hrstimef(in(Hr), go(Dest)) :: address * num + (foreign).
:- true pred get_current_hrtimef(go(Dest)) :: num + (foreign).

:- true pred get_hrtime_selff(go(Dest)) :: num + (foreign).
:- true pred get_hrvtime_selff(go(Dest)) :: num + (foreign).
:- true pred get_hrutime_selff(go(Dest)) :: num + (foreign).
:- true pred get_hrstime_selff(go(Dest)) :: num + (foreign).

% the l means that the result is long integer
:- true pred hrtime_initl(go(Error)) :: int + (foreign,returns(Error)).
:- true pred get_hrtime_structl(in(Pid), go(Dest), go(Error)) :: int * address * int + (foreign,returns(Error)).
:- true pred free_hrtime_structl(in(Hr), go(Error)) :: address * int + (foreign, returns(Error)).

:- use_foreign_library([hrtime]).
:- use_foreign_source([hrtime_c]).

:- impl_defined([
	get_hrtimef/2,
	get_hrvtimef/2,
	get_hrutimef/2,
	get_hrstimef/2,
	get_current_hrtimef/1,

	get_hrtime_selff/1,
	get_hrvtime_selff/1,
	get_hrutime_selff/1,
	get_hrstime_selff/1,

	hrtime_initl/1,
	get_hrtime_structl/3,
	free_hrtime_structl/2]).

main :-
	hrtime_initl(_E1),
	get_hrtime_structl(0,Ts,_E2),
	display(Ts),nl,
	get_hrstimef(Ts,A),
	display(A),nl,
	get_hrstimef(Ts,B),
	display(B),nl,
	free_hrtime_structl(Ts,_E3).

File: contrib/profile/hrtime/hrtime_c.c
/*
C to Prolog Adapter for the hrtime library.

Implemented by Edison Mera.

 */
#include <hrtime.h>

void get_hrtimef(struct hrtime_struct *hr, double *dest) {
  hrtime_t r;
  get_hrtime(hr, &r);
  *dest = (double)r; 
}
void get_hrvtimef(struct hrtime_struct *hr, double *dest) {
  hrtime_t r;
  get_hrvtime(hr, &r);
  *dest = (double)r; 
}
void get_hrutimef(struct hrtime_struct *hr, double *dest) {
  hrtime_t r;
  get_hrutime(hr, &r);
  *dest = (double)r; 
}
void get_hrstimef(struct hrtime_struct *hr, double *dest) {
  hrtime_t r;
  get_hrstime(hr, &r);
  *dest = (double)r; 
}
void get_current_hrtimef(double *dest) {
  hrtime_t r;
  get_current_hrtime(&r);
  *dest = (double)r; 
}

void get_hrtime_selff(double *dest) {
  hrtime_t r;
  get_hrtime_self(&r);
  *dest = (double)r;
}
void get_hrvtime_selff(double *dest) {
  hrtime_t r;
  get_hrvtime_self(&r);
  *dest = (double)r;
}
void get_hrutime_selff(double *dest) {
  hrtime_t r;
  get_hrutime_self(&r);
  *dest = (double)r;
}
void get_hrstime_selff(double *dest) {
  hrtime_t r;
  get_hrstime_self(&r);
  *dest = (double)r;
}

long hrtime_initl() {
  return hrtime_init();
}
long get_hrtime_structl(pid_t pid, struct hrtime_struct **dest) {
  return get_hrtime_struct(pid, dest);
}
long free_hrtime_structl(struct hrtime_struct *hr) {
  return free_hrtime_struct(hr);
}
